prometheus/go_collector_go117.go

```diff
@@ -0,0 +1,321 @@
+func (c *goCollector) Collect(ch chan<- Metric) {
+	// Collect base non-memory metrics.
+	c.base.Collect(ch)
+
+	// Populate runtime/metrics sample buffer.
+	metrics.Read(c.rmSampleBuf)
+
+	for i, sample := range c.rmSampleBuf {
+		desc := c.rmDescs[i]
+		switch sample.Value.Kind() {
+		case metrics.KindUint64:
+			ch <- MustNewConstMetric(desc, c.rmValTypes[i], float64(sample.Value.Uint64()))
+		case metrics.KindFloat64:
+			ch <- MustNewConstMetric(desc, c.rmValTypes[i], sample.Value.Float64())
+		case metrics.KindFloat64Histogram:
+			h := newConstRuntimeMetricsHistogram(desc, sample.Value.Float64Histogram())
+			if sum, ok := c.exactSumFor(sample.Name); ok {
+				h.exactSum(sum)
+			}
+			ch <- h
+		case metrics.KindBad:
+			// TODO: Unsupported metric. Log this somewhere?
+		default:
+			// TODO: Unsupported metric kind. Log this somewhere?
+		}
+	}
+
+	// ms is a dummy MemStats that we populate ourselves so that we can
+	// populate the old metrics from it.
+	var ms runtime.MemStats
+	memStatsFromRM(&ms, c.rmSampleMap)
+	for _, i := range c.msMetrics {
+		ch <- MustNewConstMetric(i.desc, i.valType, i.eval(&ms))
+	}
+}
+
+// exactSumFor takes a runtime/metrics metric name (that is assumed to
+// be of kind KindFloat64Histogram) and returns its exact sum and whether
+// its exact sum exists.
+//
+// The runtime/metrics API for histograms doesn't currently expose exact
+// sums, but some of the other metrics are in fact exact sums of histograms.
+func (c *goCollector) exactSumFor(rmName string) (float64, bool) {
+	switch rmName {
+	case "/gc/heap/allocs-by-size:bytes":
+		s, ok := c.rmSampleMap["/gc/heap/allocs:bytes"]
+		if !ok {
+			break
+		}
+		return float64(s.Value.Uint64()), true
+	case "/gc/heap/frees-by-size:bytes":
+		s, ok := c.rmSampleMap["/gc/heap/frees:bytes"]
+		if !ok {
+			break
+		}
+		return float64(s.Value.Uint64()), true
+	}
+	return 0, false
+}
+
+func memStatsFromRM(ms *runtime.MemStats, rm map[string]*metrics.Sample) {
+	lookupOrZero := func(name string) uint64 {
+		if s, ok := rm[name]; ok {
+			return s.Value.Uint64()
+		}
+		return 0
+	}
+
+	// Currently, MemStats adds tiny alloc count to both Mallocs AND Frees.
+	// The reason for this is because MemStats couldn't be extended at the time
+	// but there was a desire to have Mallocs at least be a little more representative,
+	// while having Mallocs - Frees still represent a live object count.
+	// Unfortunately, MemStats doesn't actually export a large allocation count,
+	// so it's impossible to pull this number out directly.
+	tinyAllocs := lookupOrZero("/gc/heap/tiny/allocs:objects")
+	ms.Mallocs = lookupOrZero("/gc/heap/allocs:objects") + tinyAllocs
+	ms.Frees = lookupOrZero("/gc/heap/frees:objects") + tinyAllocs
+
+	ms.TotalAlloc = lookupOrZero("/gc/heap/allocs:bytes")
+	ms.Sys = lookupOrZero("/memory/classes/total:bytes")
+	ms.Lookups = 0 // Already always zero.
+	ms.HeapAlloc = lookupOrZero("/memory/classes/heap/objects:bytes")
```

Comment:
all of this looks scary.. prone to errors and typos  😱

Reply:
I avoided part of the typo issue by using `MemStats` on one side of things (so the compiler at least enforces the names exist), but this is somewhat inevitable given that . I'll try to come up with some tests to help here.

I think a lot of this looks scarier than it is. These strings are going to change slowly and rarely. As long as there's testing to catch it, it should be straightforward to maintain.

Reply:
I added a test for this.